### PR TITLE
Mempool only CHECKLOCKTIMEVERIFY (BIP65) verification, unparameterized version

### PR DESCRIPTION
**Pull request by petertodd**
Same as #5496, but with without the "type" parameter. In this pull-req, a future relative CLTV or other new modes would need to be implemented with another opcode. (e.g. the original proposal)

I personally have a preference for this version, as we're [in no danger of running out of NOPs](https://github.com/bitcoin/bitcoin/pull/5496#issuecomment-101293403).

### CODE REVIEW

**Comment by btcdrak** [comment]
ACK. This version is well tested.

**Comment by NicolasDorier** [comment]
As stated, I prefer this one, it is smaller code, has been more tested, save one byte.

My aesthetic is that we will get nicer human readable script. (<locktime> OP_CLTV and <push> OP_RCLTV, instead of cryptic <locktime> <mode> OP_CLTV )

**Comment by gavinandresen** [comment]
Nit: I think CheckLockTime's full implementation would conceptually fit better implemented as part of BaseSignatureChecker (and it would remove a few lines of code), because the purpose of the BaseSignatureChecker/TransactionSignatureChecker split is just to have a checker that skips expensive ECDSA validation. Checking the locktime is fast, and it feels safer to have BaseSignatureChecker do the check.

In practice, the above nit doesn't matter, the only place BaseSignatureChecker is used at the moment is checking for non-standard scriptSigs.

Besides that, code review ACK.

**Comment by sipa** [comment]
BaseSignatureChecker does not know that the data being signed is a
transaction.

File: src/script/interpreter.cpp [hunk 1]
```diff
@@ -335,9 +335,51 @@ bool EvalScript(vector<vector<unsigned char> >& stack, const CScript& script, un
...
+                    // default 4-byte limit.
+                    //
+                    // If we kept to that limit we'd have a year 2038 problem,
+                    // even though the nLockTime field in transactions
+                    // themselves is uint32 which only becomes meaningless
+                    // after the year 2106.
+                    //
+                    // Thus as a special case we tell CScriptNum to accept up
+                    // to 5-byte bignums, which are good until 2**32-1, the
+                    // same limit as the nLockTime field itself.
```
Comment (luke-jr) [inline src/script/interpreter.cpp]: This is not clear: 5-byte CScriptNum should be good up to 2**39-1, *beyond* the limit of the nLockTime field.

Comment (petertodd) [inline src/script/interpreter.cpp]: Fixed.

**Comment by jtimon** [comment]
Yes, the whole point of BaseSignatureChecker is that EvalScript doesn't need to know anything about transactions.
Untested ACK

**Comment by NicolasDorier** [comment]
any plan to merge that for 0.11 ? I know I can play with it with viacoin, but I think there is no point in delaying it anymore since there is no controversial problems anymore with it.

**Comment by jtimon** [comment]
I don't see any reason not to do it either. maaku prefers the other version but I don't think he is opposed to this one, which seems to be preferred by almost everyone else.

**Comment by rubensayshi** [comment]
utACK

File: src/script/interpreter.h [hunk 2]
```diff
@@ -76,6 +76,10 @@ enum
     // (softfork safe, BIP62 rule 6)
     // Note: CLEANSTACK should never be used without P2SH.
     SCRIPT_VERIFY_CLEANSTACK = (1U << 8),
+
+    // Verify CHECKLOCKTIMEVERIFY (BIP65)
+    //
```
Comment (maaku) [inline src/script/interpreter.h]: Stupidest little nit... this empty comment line can be removed, no? Or better yet explain the feature like the other comments do.

**Comment by maaku** [comment]
My objections about the unparametrized version should not be interpreted as release blocking.

**Comment by petertodd** [comment]
@maaku's Fixed nit. Also thanks for the review!

**Comment by jgarzik** [comment]
Are there any remaining blockers to merging this?

**Comment by petertodd** [comment]
@jgarzik Nothing I can think of.

If for some reason we find we totally screwed up plan is to just change the behavior, use a new NOP, and go from there. (as was discussed a few months ago with no objections: https://github.com/bitcoin/bitcoin/pull/5496#issuecomment-68114665)

**Comment by jonasschnelli** [comment]
code review ACK.

Will test and write some example RPC regtests within the next days.

**Comment by petertodd** [comment]
@jonasschnelli RPC regtests? What would those regtests actually do?

I do have very simple demo's of CLTV here: https://github.com/petertodd/checklocktimeverify-demos

**Comment by jonasschnelli** [comment]
@petertodd: some simple CLTV real-world examples. Creating some CLTV transactions (needs hex fiddling because the wallet doesn't support it right now), broadcast, mine some blocks, see what's in there.
Like the other scripts in (/qa/rpc-tests).
This is a way to test a feature while also creating a test script that help to sustain quality.

**Comment by petertodd** [comment]
@jonasschnelli Hmm, I suspect the actual testing value of doing that is relatively low, at least until the soft-fork code goes in; the actual opcode itself is very well tested in the script tests, and there's no IsStandard() changes associated. (yet) I personally would add that kind of thing to [cltv-demos](https://github.com/petertodd/checklocktimeverify-demos) because I greatly prefer the python library it uses. That said, I certainly have no objection to doing that - and it might come in handy when we do add block tests for the soft-fork code - although I'd ask that we don't wait on it before merging this pull-req.

**Comment by jonasschnelli** [comment]
Sure. I didn't mention the rpc test to hold this PR back.
Merging is totally fine for me.

**Comment by jonasschnelli** [comment]
What about changing script.cppL144 from
`case OP_NOP2                   : return "OP_NOP2";`
to
`case OP_NOP2                   : return "OP_CHECKLOCKTIMEVERIFY";`
or
`case OP_CHECKLOCKTIMEVERIFY                   : return "OP_CHECKLOCKTIMEVERIFY";`

**Comment by petertodd** [comment]
@jonasschnelli Good point.

Anyone have any idea what doing that will break? :/

**Comment by btcdrak** [comment]
@petertodd the tests...

**Comment by jonasschnelli** [comment]
I came up with this because it's exposed in the RPC call `decoderawtransaction` (`scriptPubKey['asm']`).
Changing this would probably only harm users who rely on OP_NOP2 (which is not in use) in this field. If someone uses the `asm` field and parse the content he needs to get bitchslapped anyways.

**Comment by mruddy** [comment]
Nice work on this @petertodd. It'll be very useful once it's rolled out!

I ran some CLTV style micropayment channel deposit and refund tests on testnet.

Overall, the changes look good.

I'm wondering why you chose to make OP_CHECKLOCKTIMEVERIFY not pop the input argument off the top of the stack in the success case?

In the success case, by not popping, the semantic of the original no-op definition of OP_NOP2 is preserved. I assume that maintaining the no-op semantic for old clients, so that they could still validate scripts using the new semantic of the opcode, was your reason.
If so, did you have any other reason, like you were also trying to save an upgradable nopcode by not using one for a matching non-verify OP_CHECKLOCKTIME (to fit the pattern of OP_CHECKSIG+OP_CHECKSIGVERIFY, OP_EQUAL+OP_EQUALVERIFY, etc...)?

During the transition, I don't think preserving the no-op semantic in this case is desirable for the same reason that it is not desirable to start using CLTV outputs prior to the 95% lock-in during a soft fork -- the refund path in CLTV scripts could be interpreted as no-op's on non-time-locked transactions to claim the refund outputs early.

If backwards compatibility for older clients was the only reason, then I think that changing to popping the input stack argument should be considered because:
- Doing so would be consistent with the other "verify" opcodes (i.e.- OP_VERIFY, OP_EQUALVERIFY, OP_NUMEQUALVERIFY, OP_CHECKSIGVERIFY, and OP_CHECKMULTISIGVERIFY).
- Doing so is still soft-fork compatible
- During a soft-fork transition, before 95% lock-in, breaking validation of the new scripts in old clients seems preferable, since, for example, we'd like to avoid having premature adopters of CLTV bond refund scripts get early refunds going through (via finalized inputs or earlier than expected transaction nLockTimes). Having new scripts expect that the stack is popped on success would break the new scripts on older validators.
- After the semantic transition is locked-in, it seems that most of the uses of OP_CHECKLOCKTIMEVERIFY would not need the argument left on the stack. Having to repeat the OP_DROP everytime seems kludgy. People that need the argument to remain on the stack could precede the OP_CHECKLOCKTIMEVERIFY with an OP_DUP.

Make sense?

**Comment by mruddy** [comment]
Here's a table of the combinations of test cases that I tried.
The rows represent the three vintages of clients, and the columns are the two possible intended semantics of the OP_NOP2 opcode within scripts.
The cells are outcomes with the current implementation of this PR (not popping the top stack item):

 | script expected no-op semantic  | script expected cltv semantic
---|---|---
pre-SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS client (NOP2 execution treated as NOP) | script validates                | script validates (seems problematic pre-lock-in)
SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS client (NOP2 execution fails)              | script fails (NOPx reserved)    | script fails (NOPx reserved)
CLTV aware client (NOP2 execution treated as CLTV)                                        | script usually fails*           | script validates

*Note: Failures are caused by trying to get a CScriptNum from the top of the stack and by checks within TransactionSignatureChecker::CheckLockTime. Outcomes are script dependent, but for the cases I tried, always failed.

**Comment by maaku** [comment]
If you pop the argument off the stack it becomes a hard fork change.

**Comment by mruddy** [comment]
@maaku: Doh, after re-thinking, I think you are right about that.

I think I confused myself by asking the wrong question when I was contemplating whether having it pop would then be a hard vs. soft fork change.

Now, when I ask myself:
"Would the change cause a transaction to be seen as invalid by older clients, while simultaneously being seen as valid by newer clients?"

The answer is, yes. I even alluded to that in my previous message, "Having new scripts expect that the stack is popped on success would break the new scripts on older validators."

Does that match with what you were thinking? If so, then yes, please forget my suggestion of making the CLTV pop.

**Comment by maaku** [comment]
Correct.

**Comment by mruddy** [comment]
@maaku cool, thanks for correcting me on that.

@jonasschnelli and @petertodd : About the "asm" translation update, when we were talking about removing the "OP_" prefixes and adding SIGHASH decodes for #5264 and #5392, we did not come up with too much to worry about breaking when changing the "asm" results. Back then, it seemed like breaking dependencies on "asm", if they existed, was perceived as a potentially desirable outcome anyways. For example, https://github.com/bitcoin/bitcoin/pull/5264#issuecomment-63519743

**Comment by petertodd** [comment]
@mruddy Thanks for reminding me! That's a good argument. :) I'll look into doing that then.

**Comment by btcdrak** [comment]
@petertodd @mruddy Isn't that sort of out of scope for this particular PR?

**Comment by mruddy** [comment]
@btcdrak @petertodd I think updating the opcode name decode is justifiable in the interest of being complete. People do look at the "asm" in the raw transaction and script decodes. Having it read as OP_NOP2 could be misleading, especially to casual users, after the new semantic is in effect. As far as timing, it's not critical to update the opcode name decode right now (especially if it pushes this pull request out further). The decode update could be done along with the soft-fork updates, or even later as a part of an opcode decode name cleanup (where the "OP_" prefixes get dropped, etc...). But, since there is the justification of being complete with the semantic change, I think it is acceptable to break any possible ill-advised external dependencies on the existing "asm" decode.

**Comment by petertodd** [comment]
@mruddy That all sounds reasonable. After all, if changing the name isn't a big deal, then changing the name with another pull-req shouldn't be a big deal, and will cut down on the amount of unnecessary code changes in the more pull-req that does change consensus-critical code.

IMO, this pull-req is ready for merging as-is.

**Comment by mruddy** [comment]
@petertodd I'm good with that.

Tested ACK

**Comment by afk11** [comment]
utACK

**Comment by laanwj** [comment]
utACK

**Pull request merged**